core/apple/src/SinksApple.kt

```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright 2017-2023 JetBrains s.r.o. and respective authors and developers.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the LICENCE file.
+ */
+
+package kotlinx.io
+
+import kotlinx.cinterop.*
+import platform.Foundation.*
+import platform.darwin.NSInteger
+import platform.darwin.NSUInteger
+import platform.posix.uint8_tVar
+import kotlin.native.ref.WeakReference
+
+/**
+ * Returns an output stream that writes to this sink. Closing the stream will also close this sink.
+ *
+ * @sample kotlinx.io.samples.KotlinxIoSamplesApple.asStream
+ */
+public fun Sink.asNSOutputStream(): NSOutputStream = SinkNSOutputStream(this)
+
+@OptIn(UnsafeNumber::class)
+private class SinkNSOutputStream(
+    private val sink: Sink
+) : NSOutputStream(toMemory = Unit), NSStreamDelegateProtocol {
+
+    private val isClosed: () -> Boolean = when (sink) {
+        is RealSink -> sink::closed
+        is Buffer -> {
+            { false }
+        }
+    }
+
+    private var status = NSStreamStatusNotOpen
```

Comment:
If the stream is not open, `write()` will just return -1, but not change the status to error. `NSStreamStatusError` is a result of a terminal error, so any error that is correctable, like calling a function in the wrong order or with bad parameters, should not change the status to error.

I don't think we should go from `NSStreamStatusClosed` to `NSStreamStatusError`. Calling `write` on a closed platform stream just returns -1. I'll change that throwing `IOException` to returning -1 when closed.

We should probably mirror the behavior of the platform streams and not allow closing a stream that has never been opened. I'll make that change. You're missing the close from the `NSStreamStatusOpen` state though.

Where are you getting the `NSStreamStatusOpen -->|error| NSStreamStatusError;` from? I believe we should only get an `error` during `write`. That's the only place it's being set.

I guess we just pass right through `NSStreamStatusOpening`. Maybe it makes sense to set it right before setting to `NSStreamStatusOpen`, in case some key-value observer is expecting it in Objective-C? We don't actually do anything to open the underlying `Sink` or `Source`, so setting both statuses is entirely to align with `NSStream` expected behavior.

Sinks don't provide a mechanism to be limited in capacity, so we never get `NSStreamStatusAtEnd`.

This is what I believe the status lifecycle should look like after these changes:

```mermaid
graph TD
    NSStreamStatusNotOpen -->|open| NSStreamStatusOpening;
    NSStreamStatusOpening -->|open| NSStreamStatusOpen;
    NSStreamStatusOpen -->|close| NSStreamStatusClosed;
    NSStreamStatusOpen -->|write| NSStreamStatusWriting;
    NSStreamStatusWriting -->|write| NSStreamStatusOpen;

    NSStreamStatusWriting -->|error| NSStreamStatusError;

    NSStreamStatusReading;
    NSStreamStatusAtEnd;
```

Reply:
> Where are you getting the NSStreamStatusOpen -->|error| NSStreamStatusError; from? I believe we should only get an error during write. That's the only place it's being set.

That's possible when the stream is still open, but the underlying sink is closed and someone is calling `write`.
Not saying that it's wrong or problematic, just mentioning the possible transition.

> I guess we just pass right through NSStreamStatusOpening. Maybe it makes sense to set it right before setting to NSStreamStatusOpen, in case some key-value observer is expecting it in Objective-C?

My understanding is that it's mainly intended for streams with complex opening protocols, like network connections.

Reply:
> That's possible when the stream is still open, but the underlying sink is closed and someone is calling write.
Not saying that it's wrong or problematic, just mentioning the possible transition.

Oh, ok. This should be changed now where any `error` would come after `write` has changed status to `NSStreamStatusWriting`.